### PR TITLE
add MaxPriorityClassValue to psp

### PR DESCRIPTION
**Pull request by CaoShuFeng**
MaxPriorityClassValue is added to podsecuritypolicy. Cluster admin can use this field to set the max priority one pod can get.

This use case of this pull request:
Tom is a vip client of the cluster, so he can create pods with priority no larger than 2000.
While Lisa is a normal client, so she can only create pods with priority between 0~1000.

/assign @bsalamat @liggitt 

**Special notes for your reviewer**:

**Release note**:
```release-note
MaxPriorityClassValue is added to podsecuritypolicy. Cluster admin can use this field to set the max priority one pod can get.
```

### CODE REVIEW

**Comment by liggitt** [comment]
Was this part of the priority design? This seems more resource centric than security focused. What's the interaction with quota?

cc @derekwaynecarr

**Comment by CaoShuFeng** [comment]
> Was this part of the priority design?

No.  The design doc saies it's No-goal. I looks like I treate this feature too easy..

> This seems more resource centric than security focused. 
Yes, you are right. Thanks for pointing it out.

> What's the interaction with quota?

 I looks like I treate this feature too easy..
I am OK to close this pull request @liggitt .

What's the community's plan for this feature? @bsalamat 
I am very glad to help.

File: pkg/apis/extensions/validation/validation.go [hunk 1]
```diff
@@ -711,6 +712,15 @@ func validatePSPAllowedHostPaths(fldPath *field.Path, allowedHostPaths []extensi
 	return allErrs
 }
 
+// validatePSPMaxPriorityClassValue validates the MaxPriorityClassValue of PodSecurityPolicy.
+func validatePSPMaxPriorityClassValue(fldPath *field.Path, priority *int32) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if priority != nil && *priority < 0 {
```
Comment (bsalamat) [inline pkg/apis/extensions/validation/validation.go]: negative values are actually valid for priority.

File: pkg/apis/extensions/types.go [hunk 2]
```diff
@@ -865,6 +865,9 @@ type PodSecurityPolicySpec struct {
 	// is allowed in the "Volumes" field.
 	// +optional
 	AllowedFlexVolumes []AllowedFlexVolume
+	// The max integer value of priority one pod can set. PodPriority feature is required, otherwise this field is ignored.
+	// +optional
+	MaxPriorityClassValue *int32
```
Comment (bsalamat) [inline pkg/apis/extensions/types.go]: I'd suggest "MaxAllowedPriority" for the name of this field.

**Review by bsalamat — COMMENTED** [review]
This is not a bad idea, but how about system critical components which by design have very high priority? Certain pods, such as kube-dns, kube-proxy, etc have system-node-critical and system-cluster-critical priority. Such pods get very high priority which is intentional. How does your PR handle those?

**Comment by bsalamat** [comment]
> What's the community's plan for this feature? @bsalamat
I am very glad to help.

Our general idea is to support limiting priority by quota for priority classes. A user who does not have any quota for a particular priority class won't be able to create any pods with that priority.
The design doc can be [found here](https://github.com/kubernetes/community/pull/933).

**Comment by tallclair** [comment]
I agree with @liggitt . I'm not sure that PodSecurityPolicy is the right place for this.

**Comment by liggitt** [comment]
/close

let's start with the quota control over priority

**Pull request closed without merging**